azure/services/managedclusters/managedclusters.go

```diff
@@ -169,11 +169,11 @@ func (s *Service) Reconcile(ctx context.Context) error {
 	// clusters API at create time, not update.
 	if azure.ResourceNotFound(err) {
 		isCreate = true
-		// Add system agent pool to cluster spec that will be submitted to the API
-		managedClusterSpec.AgentPools, err = s.Scope.GetAgentPoolSpecs(ctx)
-		if err != nil {
-			return errors.Wrapf(err, "failed to get system agent pool specs for managed cluster %s", s.Scope.ClusterName())
-		}
+	}
```

Comment:
can you clarify the thinking here? shouldn't this still happen only on create?

if not, a code comment with the reasoning would be nice.

Reply:
I was seeing empty agentpoolprofiles on updates. I will add a comment. Without this the ManagedCluster was setting orchestrator version empty for the agentpoolprofiles and upgrading them with the managed controlplane version

I will add a comment.

Reply:
got it, thanks. 

I think we might need to be careful here -- it's possible for a user to add AMMP, and for that to cause a race between AMCP and AMMP reconcile. if AMCP wins, we'll end up adding a new agentpool via AKS ManagedClusters API which will return a 400. so you may want to do a GET and make sure you only PUT existing agentpools on update.

I guess that might cause a race in the other direction, where we do 1) GET on AKS 2) AMMP creates new pool 3) AMCP reconciles old pools and we deleted one that AMMP just created so we get a 400...

I think either way it's possible to get a 400 and kind of hard to avoid due to the upgrade issue you mentioned...I'm fine with either solution for now (likely as-is, given it's less code and no additional changes). I might have to check/ask around on the AKS side if there's a better way to handle this.

Reply:
cc @devigned in case you have some brainy ideas on avoiding this

Reply:
thanks for the review!
> make sure you only PUT existing agentpools on update

good idea. What about if there is a mismatch between GET on AKS and AMMP count/name etc, throw an error and wait till it is same? This way we let AMMP manage the creation or deletion

Reply:
Ideally if the cluster is capi managed the only mismatch should be in number of pools. I guess the pool config could change but again, we wouldn't want to modify that on cluster PUT. 

so we might need to treat 400s here as a retryable error and let re-reconcile fix it, which isn't ideal, but should work.

Reply:
I would prefer if we were able to differentiate between a 400 that is caused by an agent pool add/delete vs a different 400. 

400 is usually a client error, which requires the client to make a different request. That would mean it would generally be a terminal error, rather than transient.

Reply:
I reverted the above change. And just use later `existingMC.AgentPoolProfiles` for the PUT which is coming from `GET on AKS`

Reply:
> I would prefer if we were able to differentiate between a 400 that is caused by an agent pool add/delete vs a different 400.

I believe AKS does categorize the error in some useful way, I can check the message and get back to you.

> 400 is usually a client error, which requires the client to make a different request. That would mean it would generally be a terminal error, rather than transient.

true, but on re-reconcile we would in fact make a different request due to the race I mentioned, which would be valid. 

>I reverted the above change. And just use later existingMC.AgentPoolProfiles for the PUT which is coming from GET on AKS

unfortunately this can still hit the race in the other direction I mentioned before. I'm fine with the behavior for this PR as long as we don't consider a 400 in MC reconciler as terminal. It'd be nice to open an issue to track that separately if we have this PR go in with the logic as it is.

Reply:
> I believe AKS does categorize the error in some useful way, I can check the message and get back to you.

I will wait for this and not consider this particular error as non-terminal.
to consider it non-terminal it would be wrapping it with `azure.WithTransientError` explicitly right?